automerge-frontend/src/state_tree/sequence_tree.rs

```diff
@@ -0,0 +1,672 @@
+            })
+        }
+        assert_eq!(self.len(), old_len + 1, "{:#?}", self);
+    }
+
+    pub fn push_back(&mut self, opid: OpId, element: T) {
+        let l = self.len();
+        self.insert(l, opid, element)
+    }
+
+    pub fn get(&self, index: usize) -> Option<(OpId, &T)> {
+        self.root_node.as_ref().and_then(|n| n.get(index))
+    }
+
+    pub fn get_mut(&mut self, index: usize) -> Option<(OpId, &mut T)> {
+        self.root_node.as_mut().and_then(|n| n.get_mut(index))
+    }
+
+    pub fn remove(&mut self, index: usize) -> T {
+        if let Some(root) = self.root_node.as_mut() {
+            #[cfg(debug_assertions)]
+            let len = root.check();
+            let old = root.remove(index);
+
+            if root.elements.is_empty() {
+                if root.is_leaf() {
+                    self.root_node = None;
+                } else {
+                    self.root_node = Some(root.children.remove(0));
+                }
+            }
+
+            #[cfg(debug_assertions)]
+            debug_assert_eq!(len, self.root_node.as_ref().map_or(0, |r| r.check()) + 1);
+            old.1
+        } else {
+            panic!("remove from empty tree")
+        }
+    }
+
+    pub fn set(&mut self, index: usize, element: T) -> T {
+        self.root_node.as_mut().unwrap().set(index, element)
+    }
+}
+
+impl<T, const B: usize> SequenceTreeNode<T, B>
+where
+    T: Clone + Debug,
+{
+    fn new() -> Self {
+        Self {
+            elements: Vec::new(),
+            children: Vec::new(),
+            length: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.length
+    }
+
+    fn is_leaf(&self) -> bool {
+        self.children.is_empty()
+    }
+
+    fn is_full(&self) -> bool {
+        self.elements.len() >= 2 * B - 1
+    }
+
+    fn insert_into_non_full_node(&mut self, index: usize, opid: OpId, element: T) {
+        assert!(!self.is_full());
+        if self.is_leaf() {
+            self.length += 1;
+            self.elements.insert(index, Box::new((opid, element)));
+        } else {
+            let mut cumulative_len = 0;
+            for (child_index, child) in self.children.iter_mut().enumerate() {
+                if cumulative_len + child.len() >= index {
+                    if child.is_full() {
+                        self.split_child(child_index);
+
+                        let mut cumulative_len = 0;
+                        for child in self.children.iter_mut() {
+                            if cumulative_len + child.len() >= index {
+                                child.insert_into_non_full_node(
+                                    index - cumulative_len,
+                                    opid,
+                                    element,
+                                );
+                                self.length += 1;
+                                break;
+                            } else {
+                                cumulative_len += child.len() + 1;
```

Comment:
This `child.len() + 1` fragment pops up a lot. I feel like the code would be a little easier to read and less error prone if instead of `SequenceTreeNode::children` being a `Vec<SequenceTreeNode<..>>` it was instead a newtype wrapper with some convenience methods for this commonly used logic. Thoughts?

Reply:
I've had a little refactor and tried having a `cumulative_iterator` function but the borrow-checker gets in the way as it borrows the entire struct rather than just the field. Something I'll leave for another time I think.